rust/src/utils/scbase64.rs

```diff
@@ -0,0 +1,315 @@
+            }
+            i = &i[1..];
+        }
+        if decoder.nb == 4 {
+            decoder.tmp[0] = base64_map(decoder.tmp[0])?;
+            decoder.tmp[1] = base64_map(decoder.tmp[1])?;
+            if decoder.tmp[2] == b'=' {
+                r[offset] = (decoder.tmp[0] << 2) | (decoder.tmp[1] >> 4);
+                offset += 1;
+            } else {
+                decoder.tmp[2] = base64_map(decoder.tmp[2])?;
+                if decoder.tmp[3] == b'=' {
+                    r[offset] = (decoder.tmp[0] << 2) | (decoder.tmp[1] >> 4);
+                    r[offset + 1] = (decoder.tmp[1] << 4) | (decoder.tmp[2] >> 2);
+                    offset += 2;
+                } else {
+                    decoder.tmp[3] = base64_map(decoder.tmp[3])?;
+                    r[offset] = (decoder.tmp[0] << 2) | (decoder.tmp[1] >> 4);
+                    r[offset + 1] = (decoder.tmp[1] << 4) | (decoder.tmp[2] >> 2);
+                    r[offset + 2] = (decoder.tmp[2] << 6) | decoder.tmp[3];
+                    offset += 3;
+                }
+            }
+            decoder.nb = 0;
+        }
+        if stop {
+            break;
+        }
+    }
+    r.truncate(offset);
+    return Ok(r);
+}
+
+pub fn decode_rfc2045(decoder: &mut Decoder, input: &[u8]) -> Result<Vec<u8>> {
+    let mut i = input;
+    let maxlen = ((decoder.nb as usize + i.len()) * 3) / 4;
+    let mut r = vec![0; maxlen];
+    let mut offset = 0;
+    while !i.is_empty() {
+        while decoder.nb < 4 && !i.is_empty() {
+            if base64_map(i[0]).is_ok() || i[0] == b'=' {
+                decoder.tmp[decoder.nb as usize] = i[0];
+                decoder.nb += 1;
+            }
+            i = &i[1..];
+        }
+        if decoder.nb == 4 {
+            decoder.tmp[0] = base64_map(decoder.tmp[0])?;
+            decoder.tmp[1] = base64_map(decoder.tmp[1])?;
+            if decoder.tmp[2] == b'=' {
+                r[offset] = (decoder.tmp[0] << 2) | (decoder.tmp[1] >> 4);
+                offset += 1;
+            } else {
+                decoder.tmp[2] = base64_map(decoder.tmp[2])?;
+                if decoder.tmp[3] == b'=' {
+                    r[offset] = (decoder.tmp[0] << 2) | (decoder.tmp[1] >> 4);
+                    r[offset + 1] = (decoder.tmp[1] << 4) | (decoder.tmp[2] >> 2);
+                    offset += 2;
+                } else {
+                    decoder.tmp[3] = base64_map(decoder.tmp[3])?;
+                    r[offset] = (decoder.tmp[0] << 2) | (decoder.tmp[1] >> 4);
+                    r[offset + 1] = (decoder.tmp[1] << 4) | (decoder.tmp[2] >> 2);
+                    r[offset + 2] = (decoder.tmp[2] << 6) | decoder.tmp[3];
+                    offset += 3;
+                }
+            }
+            decoder.nb = 0;
+        }
+    }
+    r.truncate(offset);
+    return Ok(r);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_base64_decode(input: *const u8, len: u32, max_decoded: u32, mode: Base64Mode) -> *mut Base64Decoded {
```

Comment:
As metioned, naming. Should follow `SCBase64Decode` format.. But `ToCString` added or something.